Add list of physical quantities as axis names to NXdata

lukaspie opened the pull request:
Based on the [suggestions](https://github.com/FAIRmat-NFDI/nexus_definitions/issues/78#issuecomment-1780206828) by @Tommaso-Pincelli.

@domna @rettigl 

Fixes #78

domna commented on contributed_definitions/nyaml/NXmpes.yaml:
```diff
@@ -275,6 +275,54 @@ NXmpes(NXobject):
             raw(NX_NUMBER):
               doc: |
                 Raw data before calibration.
+            energy_raw(NX_NUMBER):
```
I'm not sure if we want this `..._raw` style of field names. This wouldn't be really energy values but, e.g., tof times or pixel positions for x/kx raw values. So I think on this stage here we would rather insert pixel positions and I think we should also name them such. I think this applies better to a lone NXdetector being written, because providers of a detector might not know whether this was, e.g., an energy or momentum axis.

lukaspie replied:
Ok, makes sense. I was also looking at the `NXdetector` base class and there are many fields which are similar to what we want to do, but they don't quite fit the fields that we probably need. 

I am not so familiar with the axes that are typically recorded in ARPES, maybe @rettigl and @Tommaso-Pincelli could comment. How about we start with something like this?

```yaml
(NXdetector):
  (NXdata):
    exists: recommended
    \@signal:
      enumeration: [raw]
    raw(NX_NUMBER):
      doc: |
        Raw data before calibration.
    detector_pixel_x(NX_NUMBER): # there is also x_pixel_offset in NXdetector base class
      doc: |
        Detector pixel in x direction.
      unit: NX_ANY
    detector_pixel_y(NX_NUMBER):
      doc: |
        Detector pixel in y direction.
      unit: NX_ANY
    raw_time_of_flight(NX_INT):
      # exists in NXdetector base class
      # unit: NX_PULSES
      # doc: |
      #   In DAQ clock pulses
    time_of_flight(NX_FLOAT): 
      # exists in NXdetector base class
      # doc: |
      #   Total time of flight 
      # unit: NX_TIME_OF_FLIGHT
    time_of_flight_adc(NX_FLOAT):
      doc: |
        Time-of-flight values, analog-to-digital converted.
      unit: NX_ANY
    polar_angle_manipulator(NX_FLOAT):
      unit: NX_ANGLE
      doc: |
        Polar rotation angle of the manipulator.
    azimuthal_angle_manipulator(NX_FLOAT):
      unit: NX_ANGLE
      doc: |
        Azimuthal rotation angle of the manipulator.
    delay_position(NX_FLOAT):
      doc: |
        Delay position of delay line detector.
      unit: NX_LENGTH
    delay_position_ADC(NX_FLOAT):
      doc: |
        Delay position of delay line detector, analog-to-digital converted.
      unit: NX_ANY
    voltage(NX_FLOAT):
      doc: |
        Measured detector voltage
      unit: NX_VOLTAGE
```
I am not so sure what to do about `magnetization` and `scattering_direction` since these are probably string values and not axes.

domna replied:
I think for a DLD we should have at least something for an x-position, y-position and `time_of_flight`. I don't know if there is a difference between `time_of_flight` or `time_of_flight_adc` so people might want to have the option to choose between them here. I think also `delay_position(_ADC)` would refer to the same value as time delay of the signals. We probably want also something like `delay_x` and `delay_y` for the time delay differences of a DLD as they don't really have pixels, but we might also like to allow storing the non-deconvolved delay time relative to the trigger, which would probably give us `time_delay_x` and `time_delay_y`. But @Tommaso-Pincelli and @rettigl probably know much better what is reasonable here.

For a time-integrated detector we probably just want `pixel_x` and `pixel_y` directly from the detector, right?

I was also think w.r.t. not introducing anything out of the scope of NXdetector here to not have `polar_angle_manipulator`, `azimuthal_angle_manipulator` or at least have an additional `external_AXIS`, which describes an axis which is coming from outside the detectors scope (think of a detector just being triggered for readout by the rest of the experimental setup - it would just know that it collected N images, which would flatten the external parameters to one axis, too). This we then could link with NXcalibration to the appropriate fields in the instrument and write it to the top-level NXdata, maybe we should also think of slicing axis in NXcalibration for the case that we need to rearange the dimensions of the external data axis (e.g., it is flat but two parameters where actually varied).

lukaspie replied:
Ok, there seem to be mayn open questions, let's wait what the others have to say. For now, I have just removed the raw.. notation and summarized all of the suggestions above in the app def.

domna commented on contributed_definitions/nyaml/NXmpes.yaml:
```diff
@@ -524,12 +572,46 @@ NXmpes(NXobject):
+          Polarization of the incoming beam.
+          
+          Can be any of linear polarization angle (degrees), ellipticity (arb. units).
+          Could be a link to /entry/instrument/beam/incident_polarization or
```
Can we really directly link this? afaik this is a stokes vector in these fields which wouldn't translate well to a 1D axis. But it got me thinking: How can we deal with cases where an axis value might be more complex than just a single number?

domna replied:
Also would one may collect some basic polarization settings like left-circular or right-circular

lukaspie replied:
Since @Tommaso-Pincelli gave two options of what he meant by polarization, we could also split this field into two:
```yaml
polarization_angle(NX_FLOAT):
  exists: optional
  unit: NX_ANGLE
  doc: Linear polarization angle of the incoming beam.
ellipticity(NX_FLOAT):
  exists: optional
  unit: NX_ANY
  doc: Ellipticity of the incoming beam.
```
But I agree, this doesn't help with translation from a more complex vector (in this case, the stokes vector of polarization) to a 1D axis.

domna replied:
Yes, actually using an additional ellipticity (we should also make this of unit `NX_ANGLE`) would be enough here. This is calculable from the *normalised* Stokes vector (see https://www.thorlabs.com/newgrouppage9.cfm?objectgroup_id=14200). But maybe it's better if we also replace the Stokes vector with `polarization_angle` and `ellipticity` in NXbeam (or add both?)

This would solve the problem that we cannot specify the unit category in NXbeam/polarization properly and we don't collect the intensity anymore, which we previously did with the Stokes vector. So in my view this would also be a "cleaner" representation of polarization state. However, Stokes vectors can be directly measured which might be the greater argument for keeping it. @rettigl and @Tommaso-Pincelli What are your opinions on this?

rettigl replied:
Typically, one plots these things just as function of the actually scanned parameter, i.e. the wave plate angle. Otherwise, one stokes parameter could be retrieved and used

rettigl commented:
I am a bit back and forth with this. On the one hand it might be a good idea to define some names for axes, on the other hand it makes the app-def so much longer and more cumbersome to read, and we will never include everything needed, because we cannot think of every possible experiment. What is the real benefit of doing this? To my view, it only becomes important once we define a derived appdef that makes use of these names, and then we can define the ones we need for this appdef there, no?
Also, e.g. the angular axes seem to be both too restrictive and too undefined to me. That's what we have NXtransformations for. Nor the NXmpes_arpes we introduced AXISNAME_depends attributes, which allow to store a link to an axis to a respective field the Nexus hierarchy. This allows providing some context for what a certain axis means in the context of the experiment, and to programmatically interpret the content of an axis. I suggest to include this into NXmpes, and also NXdata.

domna commented:
> I am a bit back and forth with this. On the one hand it might be a good idea to define some names for axes, on the other hand it makes the app-def so much longer and more cumbersome to read, and we will never include everything needed, because we cannot think of every possible experiment. What is the real benefit of doing this? To my view, it only becomes important once we define a derived appdef that makes use of these names, and then we can define the ones we need for this appdef there, no? Also, e.g. the angular axes seem to be both too restrictive and too undefined to me. That's what we have NXtransformations for. Nor the NXmpes_arpes we introduced AXISNAME_depends attributes, which allow to store a link to an axis to a respective field the Nexus hierarchy. This allows providing some context for what a certain axis means in the context of the experiment, and to programmatically interpret the content of an axis. I suggest to include this into NXmpes, and also NXdata.

I think it is useful for connecting the data to overarching concepts. This doesn't necessarily have to be through naming but I think it is the simplest way to realize this (this is why we came up with this as an intermediate step during our mpes discussions). That way we would be able in, e.g., nomad to directly pick the energy, momentum or other properties and extract useful information automatically (or make it available in the right places). But I agree that reading and maintaining this in the appdef is cumbersome (imagine that a new axis name is created in some sub-appdef, we would need to add this to NXmpes, too, to properly resolve these axes). Eventually, I think it would be best when these names restricted by enumerations which are only allowed to be picked by a label in a specific ontology concept. That would allow to let the set of names and concepts grow organically without cluttering the appdef and also would solve the sub-appdef problems, because in an ontology this could be easily inferred. But I think there are still a lot of discussions and refinements necessary until we will have a proper solution for this (but we decided in the previously mentioned discussion that we will work on this based on this intermediate concept).

lukaspie commented:
- added `AXISNAME_depends` to `NXdata`
- added `NXdata_mpes` with explicitly spelled axis names (glossary) -> extends `NXdata_mpes`, can only be used if we have base class inheritance
- removed named axes from NXmpes, keep energy axis -> still uses `NXdata`, to be replaced by `NXdata_mpes` in the future

I went this the base class inheritance route here because `NXdata_mpes` become too messy when I copied everything from `NXdata`.

rettigl commented on base_classes/nyaml/NXdata.yaml:
```diff
@@ -237,6 +237,21 @@ NXdata(NXobject):
          to avoid string parsing in reading applications.
+  \@AXISNAME_depends:
+    doc: |
+      Points to the path to a field defining the axis on which the ``AXISNAME`` axis depends.
```
```suggestion
      Points to the path of a field defining the axis on which the ``AXISNAME`` axis depends.
```

lukaspie replied:
Added in new commit (so that nyaml and nxdl have the same text).

rettigl commented on base_classes/nyaml/NXdata.yaml:
```diff
@@ -237,6 +237,21 @@ NXdata(NXobject):
+    doc: |
+      Points to the path to a field defining the axis on which the ``AXISNAME`` axis depends.
+
+      This concept allows to store a link to an axis to a respective field the Nexus hierarchy
```
```suggestion
      This concept allows to link an axis to a respective field in the Nexus hierarchy, thereby defining the physical quantity it represents.
```

lukaspie replied:
Added in new commit (so that nyaml and nxdl have the same text).

rettigl commented on base_classes/nyaml/NXdata.yaml:
```diff
@@ -237,6 +237,21 @@ NXdata(NXobject):
+          If a calibration has been performed, ``@AXISNAME_depends`` links to the result of
+          that calibration:
+            @AXISNAME_depends: '/entry/process/calibration/calibrated_axis'
+          If the axis is given in a detecotr, ``@AXISNAME_depends`` links to that detector axis:
+            @AXISNAME_depends: '/entry/instrument/detector/AXISNAME' for a 2D detector
```
This field does not exist in NXdetector. I think we discussed to change data in NXdetector into an NXdata group, then the following would apply:
```suggestion
          If the axis corresponds to a coordinate of a detector, ``@AXISNAME_depends`` links to that detector axis:
            @AXISNAME_depends: '/entry/instrument/detector/data/AXISNAME' for a 2D detector
```

rettigl replied:
We should also add a scanned axis:
```suggestion
          If the axis is a scanned motor, ``@AXISNAME_depends`` links to the transformation describing the respective motion, e.g.:
            @AXISNAME_depends: '/entry/instrument/detector/transformations/AXISNAME' for a motion of the detector
```

lukaspie replied:
I guess you don't want to replace the lines in red here, but rather add the green lines below. If so, I have done that.

rettigl commented on contributed_definitions/nyaml/NXmpes.yaml:
```diff
@@ -279,6 +279,88 @@ NXmpes(NXobject):
             raw(NX_NUMBER):
               doc: |
                 Raw data before calibration.
```
We should change the data field in NXdetector to NXdata, and then reference this here (i.e. named "data" in line 275).

lukaspie replied:
Done! We are however not changing the base class `NXdetector` where `data` is still of type `NX_NUMBER`.

rettigl commented on contributed_definitions/nyaml/NXmpes.yaml:
```diff
@@ -491,18 +574,18 @@ NXmpes(NXobject):
           varied axis may be for example energy, momentum, spatial coordinate, pump-probe
           delay, spin index, temperature, etc. The axes traces should be linked to the
           actual encoder position in NXinstrument or calibrated axes in NXprocess.
```
Put definitions of common axis names here into the docstring.

lukaspie replied:
Done. Sorry for the confusion, I wanted to keep the existing definitions in the `NXmpes_data` base class, but forgot to add the common axis names into the docstring. Thanks @domna for the fix!

rettigl reviewed (changes requested):
I don't understand why you moved all the explicit axis name definitions now into detector/data. I remember that we decided to remove them, and place the important ones into the docstring of NXdata

lukaspie commented:
I have updated the named axis docstrings in `NXmpes/NXdata` and  `NXmpes/NXdetector/data(NXdata)` a bit. We now have:
- A list of suggested axis names in these two docstrings (incl. suggested unit categories)
- Two new base classes:  `NXdata_mpes` and `NXdata_mpes_detector` which extend `NXdata`. We could possibly think about using them in the future in case we want to actually include named axis definitions (i.e., not just in the docstrings). As of now, they are unused.

Are we happy with this PR and can it be merged?

domna commented:
I think it's all good now, except the issue with NXmpes/.../NXdetector/data and NXdetector/data, where the latter is of the NX_NUMBER and the other of type NXdata. I'm a bit hesitant to change the old data as this might get complicated with the NIAC. Also I have the feeling we further need to refine `NXdetector` then, because there are some fields we would then write inside of the NXdata field.

Shall we name it `raw_data` in NXdetector for now and try to see if we can change this with the NIAC in the future, then we could rename `raw_data` to just `data` in the future (i.e., marking `raw_data` deprecated while pointing to `data`). @lukaspie , @rettigl What do you think?

lukaspie commented:
> I think it's all good now, except the issue with NXmpes/.../NXdetector/data and NXdetector/data, where the latter is of the NX_NUMBER and the other of type NXdata. I'm a bit hesitant to change the old data as this might get complicated with the NIAC. Also I have the feeling we further need to refine `NXdetector` then, because there are some fields we would then write inside of the NXdata field.
> 
> Shall we name it `raw_data` in NXdetector for now and try to see if we can change this with the NIAC in the future, then we could rename `raw_data` to just `data` in the future (i.e., marking `raw_data` deprecated while pointing to `data`). @lukaspie , @rettigl What do you think?

Do you also want to introduce `raw_data` in the `NXdetector` base class?

domna commented:
> > I think it's all good now, except the issue with NXmpes/.../NXdetector/data and NXdetector/data, where the latter is of the NX_NUMBER and the other of type NXdata. I'm a bit hesitant to change the old data as this might get complicated with the NIAC. Also I have the feeling we further need to refine `NXdetector` then, because there are some fields we would then write inside of the NXdata field.
> > Shall we name it `raw_data` in NXdetector for now and try to see if we can change this with the NIAC in the future, then we could rename `raw_data` to just `data` in the future (i.e., marking `raw_data` deprecated while pointing to `data`). @lukaspie , @rettigl What do you think?
> 
> Do you also want to introduce `raw_data` in the `NXdetector` base class?

I think we don't really have to and I think it's better to start the discussion to have data as NXdata instead of adding a redundant group.

lukaspie commented:
> > > I think it's all good now, except the issue with NXmpes/.../NXdetector/data and NXdetector/data, where the latter is of the NX_NUMBER and the other of type NXdata. I'm a bit hesitant to change the old data as this might get complicated with the NIAC. Also I have the feeling we further need to refine `NXdetector` then, because there are some fields we would then write inside of the NXdata field.
> > > Shall we name it `raw_data` in NXdetector for now and try to see if we can change this with the NIAC in the future, then we could rename `raw_data` to just `data` in the future (i.e., marking `raw_data` deprecated while pointing to `data`). @lukaspie , @rettigl What do you think?
> > 
> > 
> > Do you also want to introduce `raw_data` in the `NXdetector` base class?
> 
> I think we don't really have to and I think it's better to start the discussion to have data as NXdata instead of adding a redundant group.

Ok, then let's just rename `data` to `raw_data` in `NXmpes/NXdetector` for now👍

rettigl reviewed (approved):
Mostly fine, apart from some suggestions for the docstring lists

lukaspie commented on contributed_definitions/nyaml/NXmpes.yaml:
```diff
@@ -482,6 +510,27 @@ NXmpes(NXobject):
+        - **ellipticity**: Ellipticity of the incoming or outgoing beam.
+          Could be a link to /entry/instrument/beam/incident_ellipticity or
+          /entry/instrument/beam/final_ellipticity if they exist.
+          Unit category: NX_ANGLE (° or rad)
```
Intentionally did not add `sample_bias `and `temperature` since they are not particular for MPES. Kinetic_energy is already part of `energy` with `@type=kinetic`.

I am not so familiar with spin-resolved measurements, so any suggestions for the spin axis?

lukaspie commented on contributed_definitions/nyaml/NXmpes.yaml:
```diff
@@ -272,7 +272,35 @@ NXmpes(NXobject):
+              - **time_delay_x**: Time delay of delay line detector in x direction. Unit category: NX_TIME
+              - **time_delay_y**: Time delay of delay line detector in y direction. Unit category: NX_TIME
+              - **voltage**: Voltage applied to sample and sample holder. Unit category: NX_VOLTAGE
+              - **external_AXIS**: Describes an axis which is coming from outside the detectors scope.
```
I guess the plan was to store uncalibrated data here, right? Then the language from `NXmpes/data` about calibrated axes doesn't make too much sense. My proposal:
```yaml
- **pixel_x**: Detector pixel in x direction.    # This was suggested in the first workshop IIRC, so it would keep it.
- **pixel_y**: Detector pixel in y direction.
- **kx**: (Un)calibrated x axis in k-space. Unit category: NX_ANY (e.g., 1/Angström).
- **ky**: (Un)calibrated y axis in k-space. Unit category: NX_ANY (1/Angström).
- **kz**: (Un)calibrated z axis in k-space. Unit category: NX_ANY (1/Angström).
- **angular0**: Fast-axis angular coordinate (or second slow axis if angularly integrated).
  Unit category: NX_ANGLE
- **angular1**: Slow-axis angular coordinate (or second fast axis if simultaneously dispersed in 2 dimensions)
  Unit category: NX_ANGLE
- **spatial0**: Fast-axis spatial coordinate (or second slow axis if spatially integrated)
  Unit category: NX_LENGTH
- **spatial1**: Slow-axis spatial coordinate (or second fast axis if simultaneously dispersed in 2 dimensions)
  Unit category: NX_LENGTH
- **delay**: (Un)calibrated delay time. Unit category: NX_TIME (s).
- **polarization_angle**: Linear polarization angle of the incoming or
  outgoing beam.
  Unit category: NX_ANGLE (° or rad)
- **ellipticity**: Ellipticity of the incoming or outgoing beam.
  Unit category: NX_ANGLE (° or rad)              
- **time_of_flight**: Total time of flight. Unit category: NX_TIME_OF_FLIGHT
- **time_of_flight_adc**: Time-of-flight values, analog-to-digital converted.
- **external_AXIS**: Describes an axis which is coming from outside the detectors scope.
```

rettigl replied:
looks good to me.

lukaspie replied:
I have added here the (un)calibrated energy axis.

lukaspie commented:
I think this can be merged unless there is any other open questions.

We can add spin axes later if needed.

domna reviewed (approved):
Yes, we can add additional fields later. Let's merge it :)
